### PR TITLE
[AMBA] add SpongeTorch and AmbaPB support

### PR DESCRIPTION
**Pull request by linux-sdk-amba**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds SpongeTorch-based AMBA training/export support and a new AmbaPB inference backend, while improving EMA, checkpoint gating, validation, and custom export file handling. 

### 📊 Key Changes
- Added new training/export args in `default.yaml` and config validation, including `ema`, `amba_config`, `amba_chipset`, and `export_file`.
- Introduced `ultralytics/utils/amba.py` to centralize SpongeTorch integration for training, validation, export preparation, chipset selection, pruning-aware handling, and checkpoint gating.
- Updated `Trainer` to support optional SpongeTorch preparation, configurable EMA creation, gated checkpoint saving, resume synchronization, and no-EMA fallback checkpoint contents.
- Improved EMA updates for pruned models by copying pruning masks directly instead of averaging them.
- Updated `Model.export()` and `Exporter` to prepare AMBA export models, skip fusion when needed, preserve source metadata, and support custom export output paths.
- Enhanced `Validator` to better handle live-model validation, compiled models, dtype/autocast behavior, unwrap loss access, and AMBA-aware model resolution.
- Added `AmbaPBBackend` and registered it in `AutoBackend` for Ambarella PB inference, including model loading, CVFlow execution, metadata loading, output formatting, and backend detection for `*.ambapb.ckpt.onnx` files.
- Extended backend registration/export flow to avoid unsupported fusion/device assumptions for AMBA models.

### 🎯 Purpose & Impact
- Enables AMBA-focused workflows across training, validation, export, and inference within the Ultralytics package.
- Improves reliability when working with pruned or SpongeTorch-prepared models by avoiding unsafe fusion, preserving correct mask behavior, and aligning EMA/checkpoint state with reparameterized models.
- Gives users more control over export destinations and validation behavior for specialized deployment pipelines.
- Expands edge and embedded deployment support with native Ambarella PB backend integration.
<details><summary>📋 Skipped 2 files (lock files, generated, images, etc.)</summary>

- `ultralytics/assets/bus.jpg`
- `ultralytics/assets/zidane.jpg`
</details>

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @linux-sdk-amba, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help streamline review; an engineer will assist shortly. For this AMBA, SpongeTorch, and export/backend update, please review the checklist below 😊

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/engine/trainer.py [hunk 1]
```diff
@@ -628,24 +634,48 @@ def _model_train(self):
...
+            if not all(torch.isfinite(v).all() for v in ema.state_dict().values() if isinstance(v, torch.Tensor)):
+                LOGGER.warning(f"Skipping checkpoint save at epoch {self.epoch}: EMA contains NaN/Inf")
+                return False
+            updates = self.ema.updates
+            model_ckpt = None  # final/resume checkpoints derive from EMA when available
+        else:
+            ema, updates = None, None
+            # No EMA: persist the live training model so downstream tooling
+            # (strip_optimizer, load_checkpoint, exporters) has weights to load.
+            model_ckpt = deepcopy(unwrap_model(self.model)).half()
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py:668]: ⚠️ **HIGH**: When `ema=False`, checkpointing falls back to `deepcopy(unwrap_model(self.model))`. That reintroduces the same deepcopy problem you already worked around in `make_training_ema()` for pruned SpongeTorch modules (`weight` views backed by `weight_orig/_mask`). A no-EMA AMBA run can fail at save time or write an inconsistent checkpoint. Please reuse the same safe-copy path here or rebuild from `state_dict()` instead of deepcopying the live pruned module.

File: ultralytics/utils/amba.py [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+    Applies spongetorch ``prepare()`` (or chipset selection) when validating a ``.pt``
+    checkpoint under an amba config, and disables Conv+BN fusion for amba/pruned models.
+
+    Returns:
+        (Any): Model reference to pass to ``AutoBackend`` (path or prepared module).
+        (bool): Whether AutoBackend should fuse Conv+BN layers.
+    """
+    amba_mode = bool(getattr(args, "amba_config", None) or getattr(args, "amba_chipset", None))
+    is_pruned = isinstance(model_ref, nn.Module) and is_pruned_model(model_ref)
+    if amba_mode and isinstance(model_ref, (str, Path)) and str(model_ref).endswith(".pt"):
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/amba.py:165]: ⚠️ **HIGH**: This only loads `.pt` checkpoints to detect pruning when `amba_config`/`amba_chipset` is set. If a user validates an already-compressed checkpoint as plain `model=pruned.pt`, `is_pruned` stays `False` and the function returns `fuse=True`, so `AutoBackend` can still Conv+BN-fuse a pruned model. That makes validation dependent on redundantly passing AMBA args and can corrupt the sparse structure for standalone `.pt` inference/val. Please detect pruning for `.pt` paths regardless of `amba_mode` before deciding the `fuse` flag.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The AMBA/SpongeTorch integration is thoughtfully structured, and the new backend/export plumbing looks cohesive. I found two functional issues to address before merge: no-EMA checkpoint saving still deepcopies pruned models unsafely, and validation only disables fusion for pruned `.pt` checkpoints when AMBA args are explicitly passed.

💬 Posted 2 inline comments

<details><summary>📋 Skipped 2 files (lock files, generated, images, etc.)</summary>

- `ultralytics/assets/bus.jpg`
- `ultralytics/assets/zidane.jpg`
</details>
